a new version of subscribe implementation

adibrastegarnia opened the pull request.

Commit:
a new version of subscribe implementation

Andrea-Campanella commented on pkg/gnmi/subscribe.go:
```diff
@@ -16,12 +16,290 @@
+	if err := s.checkEncodingAndModel(req.GetSubscribe().GetEncoding(), req.GetSubscribe().UseModels); err != nil {
+		return status.Error(codes.Unimplemented, err.Error())
+	}
+	errChan := make(chan error)
```
who is actually reading from this channel ?

adibrastegarnia replied:
Andrea, 

Nothing is finalized yet and I am exploring the usage of an error channel as well to return the error from other functions to the subscribe function. However, its usage is a little bit tricky because It is not obvious to me how interacts with response channel and goroutines which leads to a deadlock sometimes. I am exploring some possibilities to come up with the best one.

adibrastegarnia commented:
@Andrea-Campanella 
It is ready for review. The new version supports everything that we had before but the code is more clean and optimized. There are still room to make it better but I think it is good enough at this stage to be merged. Please try subscribe once, poll, stream and also changing the state variable randomly and let me know if you have any comments that I can address. After merging this, I will get back to work on handling data type for get operation.

Andrea-Campanella commented on pkg/gnmi/set.go:
```diff
@@ -232,9 +232,23 @@ func (s *Server) Set(ctx context.Context, req *pb.SetRequest) (*pb.SetResponse,
+			Path: response.GetPath(),
+		}
+
+		select {
```
247 to 251 coudl probably be simplified just to `s.ConfigUpdate <- update:`

Andrea-Campanella commented on pkg/gnmi/subscribe.go:
```diff
@@ -16,12 +16,80 @@
-// Subscribe method is not implemented.
+// processSubscribeOnce processes subscribe once requests
+func (s *Server) processSubscribeOnce(c *streamClient,
+	request *pb.SubscriptionList) {
```
no need for another line, valid for different methods in this class

Andrea-Campanella commented on pkg/gnmi/util.go:
```diff
@@ -324,3 +333,242 @@ func setPathWithoutAttribute(op pb.UpdateResult_Operation, curNode map[string]in
+		}
+
+		if err == nil {
+			select {
```
can be simplified to `c.UpdateChan <- update`

Andrea-Campanella commented on pkg/gnmi/util.go:
```diff
@@ -324,3 +333,242 @@ func setPathWithoutAttribute(op pb.UpdateResult_Operation, curNode map[string]in
+}
+
+// buildSubResponse builds a subscribeResponse based on the given Update message.
+func buildSubResponse(update *pb.Update) (*pb.SubscribeResponse, error) {
```
please see the update https://github.com/onosproject/onos-config/pull/373 in ONOS-config to handle deletes

Andrea-Campanella reviewed (changes requested):
I like the new approach, some elements need updating though. Also as a general suggestion moving tons of code around is difficult to understand for reviewers, I agree that this couldn't be avoided this time.

adibrastegarnia commented:
> I like the new approach, some elements need updating though. Also as a general suggestion moving tons of code around is difficult to understand for reviewers, I agree that this couldn't be avoided this time.
@Andrea-Campanella 
Thanks for your comments Andrea. I agree with your comment but as you said I didn't have choice this time. I will take care of other comments and then update you soon.

adibrastegarnia commented:
@Andrea-Campanella 

I think I addressed your comments and it is ready for review. Please check the delete part also. 

P.S. gnmi cli uses different print functions for different subscriptions so please consider that.
